### DISCUSSION TITLE
`pnpm install --force` how to ignore the optionalDependencies  don't satisfy the current environment(cpu, os, arch)

### DISCUSSION THREAD

**Discussion by tjx666** [pnpm/pnpm#6579] [Q&A]
In [documentation](https://pnpm.io/cli/install#--force):

> --force
Force reinstall dependencies: refetch packages modified in store, recreate a lockfile and/or modules directory created by a non-compatible version of pnpm. Install all optionalDependencies even they don't satisfy the current environment(cpu, os, arch).

I just want to refetch packages modified in store, recreate a lockfile.

**Comment by benquarmby** [comment]
TL;DR: This is the only workaround I know of right now:
```sh
# Do all the desirable things without any optional dependencies at all (compatible or incompatible)
pnpm i --force --no-optional
# Install _compatible_ optional dependencies
pnpm i
```

We are in the same boat. While it would be ideal if PNPM could always assume nothing other than itself ever touched `node_modules` and the store, the reality is that outside mutations, tampering, stale file systems, etc. happen regularly. I won't list all the possible scenarios because there are too many.

There needs to be a safe (i.e. idempotent) way get out of trouble and put everything back in the pristine PNPM way _without unwanted side effects_. From my experience, incompatible optional dependencies (for example those designed only for Linux, but will still run on Mac) can cause even more chaos than a store with non-PNPM mutations.

- ✅ refetch packages modified in store,
- ✅ recreate a lockfile and/or
- ✅ recreate modules directory
- ❌ Install all optionalDependencies even they don't satisfy the current environment (cpu, os, arch).

I would love to see something like `pnpm i --rebuild` or `pnpm i --pristine` that has the same exact outcome as `pnpm i` on the happy path (this is what I meant by idempotent). There would be no need for the "I hope you know what you're doing" warning.

Edit: See also https://github.com/orgs/pnpm/discussions/3175
